### PR TITLE
Lazy Diagnostic storage allocation

### LINKED ISSUE #10727: `DiagnosticStorage` is being allocated eagerly for every IR
For a (very) simple Enso program we seem to allocate a lot of `DiagnosticStorage` objects:
Screenshot from 2024-07-31 15-49-45

The allocation could be done lazily as typically we need a very small fraction of those.

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

close #10727

Changelog:
- add: implement lazy allocation of `DiagnosticStorage` objects
- update: compiler phases to make sure they don't allocate redundant `DiagnosticStorage` objects
- update: cleanup `DiagnosticStorage` class

### Important Notes

As a result, all 10MB of redundant `DiagnosticStorage` allocations are gone

#### Before
2024-08-21-170148_2092x171_scrot

#### After
2024-08-21-170058_1693x131_scrot

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

File: engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java [hunk 1]
```diff
@@ -103,6 +104,27 @@ default String pretty() {
...
+   * @return the copy of diagnostic storage
+   */
+  DiagnosticStorage diagnosticsCopy();
+
+  /**
+   * Get storage for compiler diagnostics associated with this IR node.
+   *
+   * @return the diagnostic storage of this node
+   */
+  DiagnosticStorage getDiagnostics();
```
Comment (JaroslavTulach) [inline engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java:115]: There already is `diagnostics()`. Why do we need another `getDiagnostics()`?

File: engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java [hunk 2]
```diff
@@ -103,6 +104,27 @@ default String pretty() {
...
+   * @return the list of diagnostics
+   */
+  List<Diagnostic> diagnosticsList();
+
+  /**
+   * Get the copy of diagnostic storage associated with this IR node.
+   *
+   * @return the copy of diagnostic storage
+   */
+  DiagnosticStorage diagnosticsCopy();
```
Comment (JaroslavTulach) [inline engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java]: Why don't we just call `diagnostics().copy()`?

File: engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java [hunk 3]
```diff
@@ -103,6 +104,27 @@ default String pretty() {
   /** Storage for compiler diagnostics related to the IR node. */
   DiagnosticStorage diagnostics();
 
+  /**
+   * Get all diagnostic info associated with this IR node.
+   *
+   * @return the list of diagnostics
+   */
+  List<Diagnostic> diagnosticsList();
```
Comment (JaroslavTulach) [inline engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java]: Why don't we call `diagnostics().toList()`?

**Review by JaroslavTulach — CHANGES_REQUESTED** [review]
I am not really happy to see four diagnostics related methods in `IR`. I'd like to hear some justification before I even try to look at the massive changes around. Can you somehow update Javadoc of the diagnostics related methods in `IR` to describe why there have to be four and what's different among each other?

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+trait LazyDiagnosticStorage { self: IR =>
+
+  private[this] var _diagnostics: DiagnosticStorage = _
+
+  protected def diagnostics_=(diagnostics: DiagnosticStorage): Unit = {
+    assert(_diagnostics eq null)
+    _diagnostics = diagnostics
+  }
+
+  override def diagnostics: DiagnosticStorage = {
```
Comment (JaroslavTulach) [inline engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala:14]: I am not really sure what `_diagnostics: DiagnosticStorage = _`, but if I parse the ideas in this file, I would really prefer to get only:
```java
interface IR  {
  DiagnosticsStorage diagnostics(boolean canReturnNull);
}
```
the implementation in `LazyDiagnosticsStorage` would be:
```scala
trait LazyDiagnosticStorage { self: IR =>
  private[this] var d: DiagnosticsStorage = null;
  override synchronized def diagnostics(boolean canReturnNull) = {
    if (!canReturnNull && d == null) {
      d = DiagnosticsStorage();
   }
   d
  }
}
```
That's all we need to defer allocation of `DiagnosticsStorage` until needed, in my opinion. The boilerplate of `diagnosticsList`, `getDiagnostics`, `diagnosticsCopy` can be moved elsewhere or preferably disappear.

File: engine/runtime-integration-tests/src/test/scala/org/enso/compiler/test/core/ir/DiagnosticStorageTest.scala [hunk 5]
```diff
@@ -42,115 +41,5 @@ class DiagnosticStorageTest extends CompilerTest {
       diagnostics.toList should contain(mkDiagnostic("b"))
       diagnostics.toList should contain(mkDiagnostic("c"))
     }
-
-    "mapping across the diagnostics to produce a new sequence" in {
```
Comment (Akirathan) [inline engine/runtime-integration-tests/src/test/scala/org/enso/compiler/test/core/ir/DiagnosticStorageTest.scala:46]: What is the reason for removing all these tests?

File: engine/runtime-integration-tests/src/test/scala/org/enso/compiler/test/CompilerTest.scala [hunk 6]
```diff
@@ -151,23 +151,23 @@ trait CompilerRunner {
       */
     def asMethod: definition.Method = {
       new definition.Method.Explicit(
-        Name.MethodReference(
```
Comment (Akirathan) [inline engine/runtime-integration-tests/src/test/scala/org/enso/compiler/test/CompilerTest.scala:154]: How is this change related to diagnostic storage?

File: engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java (hunk 2 above)
Comment (hubertp) [reply engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java]: I think this makes more sense if you put it into the context of `LazyDiagnosticStorage`, which deal with the possibility of `null`.
Although I would probably already provide a default implementation here:
```
DiagnosticStorage diagnosticCopy() {
  if (diagnostics == null) null else diagnostics.copy();
}
```
Also I would maybe rename it to `diagnosticsLazyCopy`

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala (hunk 4 above)
Comment (hubertp) [reply engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala:14]:
```
_diagnostics: DiagnosticStorage = _
```
represents the actual allocated `DiagnosticStorage` and `_` means null in this context. You call it `d` he calls it `_diagnostics` which is more in line with Scala's way of defining custom setters/getters. I think this is fine.
Also you would have to `canReturnNull` all over the place and I don't think this is any better.

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    _diagnostics
+  }
+
+  override def diagnosticsList: List[Diagnostic] = {
```
Comment (hubertp) [inline engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala]: Oh, now I get it. You essentially have all those methods to deal with the possibility of `null` and not repeat the logic all over the place.

**Review by hubertp — APPROVED** [review]

File: engine/runtime-integration-tests/src/test/scala/org/enso/compiler/test/pass/analyse/DataflowAnalysisTest.scala [hunk 8]
```diff
@@ -356,31 +356,31 @@ class DataflowAnalysisTest extends CompilerTest {
     val plusSymbol       = mkDynamicDep("+")
 
     // The Identifiers
-    val methodId           = mkStaticDep(method.getId())
```
Comment (hubertp) [inline engine/runtime-integration-tests/src/test/scala/org/enso/compiler/test/pass/analyse/DataflowAnalysisTest.scala]: Please leave those with brackets. `getId` may involve side-effects (as in, actually inferring the new `id`) and as such should require brackets.

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala (hunk 4 above)
Comment (4e6) [reply engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala:14]: With this approach
```scala
ir.diagnostics(canReturnNull = true).copy()
```
can throw NPE, that's why there is a to DRY the null checking logic
```scala
ir.diagnosticsCopy()
```

Comment (4e6) [reply engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala:14]: I think I can remove the extra methods from the `IR` interface and make them more implementation-specific.

File: engine/runtime-integration-tests/src/test/scala/org/enso/compiler/test/core/ir/DiagnosticStorageTest.scala (hunk 5 above)
Comment (4e6) [reply engine/runtime-integration-tests/src/test/scala/org/enso/compiler/test/core/ir/DiagnosticStorageTest.scala:46]: They were testing the logic that was merely delegating to the underlying list of DiagnosticStorage. This logic is no longer used

File: engine/runtime-integration-tests/src/test/scala/org/enso/compiler/test/CompilerTest.scala (hunk 6 above)
Comment (4e6) [reply engine/runtime-integration-tests/src/test/scala/org/enso/compiler/test/CompilerTest.scala:154]: During the compilation, the `Explicit` node is always created from the `Binding` copying the accompanying diagnostic, location, etc. The constructor was updated to DRY it.

File: engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java (hunk 1 above)
Comment (4e6) [reply engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java:115]: One can return null. I updated the docs

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala (hunk 4 above)
Comment (JaroslavTulach) [reply engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala:14]: > With this approach
> 
> ```scala
> ir.diagnostics(canReturnNull = true).copy()
> ```
> 
> can throw NPE, that's why there is a to DRY the null checking logic

In order to be DRY for internal stuff you duplicate a single _publicly visible_ method four times! That's not DRY! To be DRY you'd write a helper method somewhere (hidden enough from regular users of the `IR`) and used it in all the internal implementations to avoid NPE:
```java
public static DiagnosticsStorage copyOrNull(DiagnosticsStorage sOrNull) {
  return sOrNull == null ? null : sOrNull.copy();
}
```
and then called it as `DiagnosticsStorage.copyOrNull(ir.diagnostics(true))`.

> You call it d he calls it

I don't care about any of the proposed names. I just believe four methods related to diagnostics in `IR` is 75% too much. I believe single `DiagnosticsStorage diagnotics(boolean canReturnNull)` would be enough to satisfy all use-cases and keep [conceptual surface](http://wiki.apidesign.org/wiki/Conceptual_surface) (some may call it DRY) smaller.

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala (hunk 7 above)
Comment (JaroslavTulach) [reply engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala]: I believe so. Thus in order to simplify **SPI** (e.g. implementations re-written by Dmitry in this PR), we have complicated [Client API](http://wiki.apidesign.org/wiki/ClientAPI). I see it as a completely opposite approach than the right one - the **usage of `IR` should be simple**. The implementation needn't be - API authors have to live thru their pain.

File: engine/runtime-integration-tests/src/test/scala/org/enso/compiler/test/pass/analyse/DataflowAnalysisTest.scala (hunk 8 above)
Comment (JaroslavTulach) [reply engine/runtime-integration-tests/src/test/scala/org/enso/compiler/test/pass/analyse/DataflowAnalysisTest.scala]: When talking about `getId()` - there used to be a PR that saved some memory by not allocating `UUID`s. Have that PR need for different versions of `getId()`, @hubertp? If not, then it might be an example to follow when making `DiagnosticStorage` lazy.

File: engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java [hunk 9]
```diff
@@ -100,9 +100,20 @@ default String pretty() {
...
+   * Get the storage for compiler diagnostics related to the IR node.
+   *
+   * @return the diagnostic storage or {@code null} if not initialized
+   */
   DiagnosticStorage diagnostics();
 
+  /**
+   * Initialize and get the storage for compiler diagnostics associated with this IR node.
+   *
+   * @return the diagnostic storage of this node
```
Comment (JaroslavTulach) [inline engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java]:
```suggestion
   * @return the diagnostic storage of this node. The result is never {@code null}
```

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  protected def diagnostics_=(diagnostics: DiagnosticStorage): Unit = {
+    assert(_diagnostics eq null)
+    _diagnostics = diagnostics
+  }
+
+  override def diagnostics: DiagnosticStorage = {
+    _diagnostics
+  }
+
+  def diagnosticsList: List[Diagnostic] = {
```
Comment (JaroslavTulach) [inline engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala]: Make `protected`

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+    if (_diagnostics eq null) Nil else _diagnostics.toList
+  }
+
+  def diagnosticsCopy: DiagnosticStorage = {
```
Comment (JaroslavTulach) [inline engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala:18]: Can it also be `protected`?

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala [hunk 12]
```diff
@@ -0,0 +1,32 @@
+package org.enso.compiler.core.ir
+
+import org.enso.compiler.core.IR
+
+trait LazyDiagnosticStorage { self: IR =>
+
+  private[this] var _diagnostics: DiagnosticStorage = _
+
+  protected def diagnostics_=(diagnostics: DiagnosticStorage): Unit = {
```
Comment (JaroslavTulach) [inline engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala:9]: Who's calling setter (this is setter, right)? There was no setter before, so I don't think we need a method for it now.

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    if (_diagnostics eq null) Nil else _diagnostics.toList
+  }
+
+  def diagnosticsCopy: DiagnosticStorage = {
+    if (_diagnostics eq null) _diagnostics else _diagnostics.copy
+  }
+
+  override def getDiagnostics: DiagnosticStorage = {
+    if (_diagnostics eq null) {
+      _diagnostics = new DiagnosticStorage()
```
Comment (JaroslavTulach) [inline engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala:24]: This is not ready for multi threaded access. Are we OK with that?

**Review by JaroslavTulach — DISMISSED** [review]
Looks like `diagnosticsCopy` and `diagnosticsList` are gone from `IR` and are only in `LazyDiagnosticsStorage` - more hidden, but still visible. I believe those methods shouldn't be visible at all. Moreover the initialization of `_diagnostics` field isn't thread safe. So I don't believe I can approve this change - but I have "requested changes"...

Approval just to _un-"request changes". Dmitry please find approval from someone else, don't count mine as +1.

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala (hunk 13 above)
Comment (4e6) [reply engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala:24]: The entire compilation logic is not thread-safe. Compilation jobs ensure that the process is single-threaded

**Comment by 4e6** [comment]
@JaroslavTulach GitHub allows reviewer to dismiss their review 
2024-08-23-164035_1421x575_scrot

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala (hunk 12 above)
Comment (4e6) [reply engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala:9]: We need a setter when creating an IR node and want to pass some existing diagnostic storage.

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala (hunk 11 above)
Comment (4e6) [reply engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala:18]: No, it is used in compiler passes.

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala (hunk 13 above)
Comment (JaroslavTulach) [reply engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala:24]: Why are we then using immutable `IR` and not modifying `IR` in place? Ó the wonders of _like to be_ functional programming!
